### ISSUE TITLE
Generate TOC when exporting to HTML/PDF

### ISSUE THREAD

**Issue by joe223** [marktext/marktext#80]
### A new feature：TOC

If there is a way to generator TOC automatically. Or,  I just type `[toc]` in the first line，it will turn to A catalog and update while content is changing.

**Comment by joe223** [comment]
weird, I can not add a Label.

**Comment by Jocs** [comment]
@joe223 Do you mean to write HTML directly in Mark Text?

**Comment by joe223** [comment]
@Jocs  just write `[TOC]`.

example markdown content:
```markdown
[toc]

# Chapter 1

## Chapter 1.1

## Chapter 1.2

# Chapter 2
```
finally, we got

```html
<ul class="toc">
    <li><a href="#1-1">Chapter 1</a></li>
        <ul class="toc">
            <li><a href="#1-1">Chapter 1-1</a></li>
            <li><a href="#1-1">Chapter 1-2</a></li>
        </ul>
    <li><a href="#1-1">Chapter 2</a></li>
</ul>

# Chapter 1

## Chapter 1.1

## Chapter 1.2

# Chapter 2
```

**Comment by Jocs** [comment]
>weird, I can not add a Label.
What's the meaning?

**Comment by joe223** [comment]
![image](https://user-images.githubusercontent.com/10026019/37979591-8b4c30e8-321b-11e8-8320-5532cf74ccad.png)

Only owner can set this.

**Comment by khaosdoctor** [comment]
I think there's an easier way to set a TOC within the text editor, like the [VSCode extension](https://marketplace.visualstudio.com/items?itemName=joffreykern.markdown-toc) does. Instead of adding a HTML string, just add more Markdown stuff inside some marked comments.

**Comment by gitfineon** [comment]
TOC is not defined in GFM, so I suggest to only add this as export option. Because how do you want to keep the table of contents up-to-date if there is already one inserted. You have to delete the old one and insert it again.

An export option for instance by replacing a special string like `%TOC-H3%` with the markdown representation of all headlines down to level 3 would be easier. A live representation is already there, so we don't need it twice.

But this is a non-critical feature, it does not make Mark Text more stable. Why is it selected for [ A more stable Mark Text ](https://github.com/marktext/marktext/projects/4). How about adding adding a Feature Wishlist project for the fun stuff.

**Comment by Jocs** [comment]
Why hasn't the TOC function been implemented?
Because I haven’t discovered the TOC in other places so far, I have not add it to Mark Text.
As you said, it does not belong to  [A more stable Mark Text ](https://github.com/marktext/marktext/projects/4).

**Comment by Jocs** [comment]
@fxha Yes, it is not implemented, because TOC is not the syntax in the GFM standard, and we also implemented TOC navigation in the sidebar, so I thought it was not necessary to implement it. When exporting, I think we can add an option. Whether to export the TOC. This is a good idea, maybe add it later when optimizing the export feature.

**Comment by Dorian-DS** [comment]
> @fxha Yes, it is not implemented, because TOC is not the syntax in the GFM standard, and we also implemented TOC navigation in the sidebar, so I thought it was not necessary to implement it. When exporting, I think we can add an option. Whether to export the TOC. This is a good idea, maybe add it later when optimizing the export feature.

The GFM should be considered minimal requirements, not maximum allowed.  Additional features will not harm its use.  There are other additional features in Marktext that are not part of GFM so I don't see why this can't be an editor feature.  The sidebar navigation feels very much like an addon and removes the clean interface of seeing only the document itself how the author intended.

**Comment by DutchPete** [comment]
> The sidebar navigation feels very much like an addon and removes the clean interface of seeing only the document itself how the author intended.
>
I certainly do NOT agree with that, and it definitely does NOT remove the clean interface: the width can be set by the user. The sidebar navigation is a very useful feature and the devs have done a very good job implementing it.

What's more, for those who do not like the side bar or temporarily do not need it, they can turn it off in the menu, *View*, *Toggle Side Bar*.

**Comment by dakira** [comment]
@Jocs There doesn't need to be any extra syntax. Just ask if a TOC should be included on export. stackedit.io allows for this:

<img width="454" alt="Bildschirmfoto 2019-11-04 um 21 29 17" src="https://user-images.githubusercontent.com/576555/68155545-34cca380-ff4a-11e9-84d6-545ac9a7a1e1.png">

The result adds a navigational sidebar to the html like this:

<img width="431" alt="Bildschirmfoto 2019-11-04 um 21 33 50" src="https://user-images.githubusercontent.com/576555/68155878-e8359800-ff4a-11e9-8cf0-df1f051bc59a.png">

**Comment by tweinreich** [comment]
> > @fxha Yes, it is not implemented, because TOC is not the syntax in the GFM standard, and we also implemented TOC navigation in the sidebar, so I thought it was not necessary to implement it. When exporting, I think we can add an option. Whether to export the TOC. This is a good idea, maybe add it later when optimizing the export feature.
> 
> The GFM should be considered minimal requirements, not maximum allowed. Additional features will not harm its use. There are other additional features in Marktext that are not part of GFM so I don't see why this can't be an editor feature. The sidebar navigation feels very much like an addon and removes the clean interface of seeing only the document itself how the author intended. Many documents are created using [TOC] which is supported on other editors.

I also hope it doesn't define the maximum that is possible. A `[TOC]` that is visible while editing, updates itself and also appears in an export would be really nice (_I don't like the sidebar opened while writing as well_).

**Comment by kataras** [comment]
Hello,

This feature is not yet implemented? I am trying for hours to find a way to generate a TOC here...

**Comment by fxha** [comment]
I already have a basic implementation for this locally and can create a draft PR in the following days. It's not feature rich but works for navigating the file.

**Comment by CrystalRays** [comment]
Could Mark Text export pdf with the index?
What I mean is not the content table in document but the index info that could be recognized by the pdf viewer which will show it on the side bar.

**Comment by kouichi-c-nakamura** [comment]
Please give us a choice in Preference to enanle [toc] on Editor. It's simply a pain to go deal with a big markdown without [toc] jumps. Otherwise I may have to find another alternative...

**Comment by cmsax** [comment]
Sidebar with `table of contents` is already there, why not use it?

**Comment by qds-soft** [comment]
GitLab has [TOC] implemented now. Would be really helpful to see it implemented in marktext, since I'm writeng quite extensive documentations for my projects.
I'd love to see it. Otherwise marktext is a really awesome editor.

**Comment by willzhang** [comment]
WE NEED IT ,THANKS

**Comment by SekoiaTree** [comment]
I dug through the code a bit; even though the implementation for exports is "there", it seems that the option is disabled. Is that intentional?
https://github.com/marktext/marktext/blob/develop/src/renderer/util/pdf.js#L93 returns nothing if `tocList` is empty, and that method is called by
https://github.com/marktext/marktext/blob/develop/src/renderer/util/pdf.js#L120 where the parameter `toc` is passed to generateHtmlToc.
This is then called by https://github.com/marktext/marktext/blob/develop/src/renderer/components/editorWithTabs/editor.vue#L942, which passes `this.editor.toc`, which is created by https://github.com/marktext/marktext/blob/develop/src/renderer/components/editorWithTabs/editor.vue#L552, which does not have any "toc" object. Therefore, the table of contents is never generated.

Several places also use https://github.com/marktext/marktext/blob/2bb405a03ee52cfba2afe53da4656abb799a0727/src/muya/lib/utils/exportHtml.js#L133 with no toc parameter passed, so maybe that's also the cause?

**Comment by 2catycm** [comment]
Hello,

This feature is not yet implemented? I am trying for hours to find a way to generate a TOC too.

**Comment by DutchPete** [comment]
@allenzt have a look at this: https://github.com/marktext/marktext/issues/1290#issuecomment-1172914762

**Comment by Jocs** [comment]
Thank you for this request. Both of the features discussed in this thread are now implemented in MarkText:

**TOC navigation sidebar** — a live, collapsible table of contents panel in the side bar tracks headings as you edit (View → Toggle Side Bar).

**TOC in HTML/PDF export** — if your document contains a standalone `[TOC]` line, it is replaced with a fully linked table of contents in the exported HTML/PDF. The export settings dialog lets you control whether the top-level heading is included and set a custom title for the TOC section.

These features are backed by the `muya.getTOC()` API in the new editor engine (`packages/muya/src/state/getTOC.ts`) and wired into the export pipeline in `packages/desktop/src/renderer/src/util/exportHtml.ts` and `pdf.ts`.

Closing as already implemented. If you encounter a specific gap or regression, please open a new issue with reproduction steps and the MarkText version.

**Comment by Ken-Leo** [comment]
Even so, the exported PDF files DO NOT have page numbering in the TOC and just dots. Would this be addressed in the future?
